WIP | Refactor kimera_vio_ros port

ruffsl opened the pull request.

ruffsl committed:
Checkin inital cmake package

ruffsl committed:
Clean up dependencies

ruffsl committed:
Update .gitignore

ruffsl committed:
Checkin inital base interface

ruffsl committed:
Return boost as depend

ruffsl committed:
Use shared pointers

ruffsl committed:
Set BaseInterface to inherit DataProviderInterface
to access protected imu_single_callback_

ruffsl committed:
Add ImuInterface

ruffsl committed:
ament uncrustify

ruffsl committed:
Add inital StereoInterface

ruffsl committed:
Use toCvShare

ruffsl committed:
Correct header ifndef block

ruffsl committed:
Refactor StereoInterface to use ImageInterface
So that any future MonoInterface may leverage the same code

ruffsl committed:
Stage stereo_vio node

ruffsl committed:
Use dynamic_cast to avoid ambiguous overload
TODO: Think of a less hacky workaround

ruffsl committed:
Add optional node_options

ruffsl committed:
Add stereo_vio_node executable

ruffsl committed:
Use gflags but drop glog for now

ToniRV commented on kimera_vio_ros/src/interfaces/stereo_interface.cpp:
```diff
@@ -9,24 +9,23 @@ namespace interfaces
   try {
-    // TODO(Toni): here we should consider using toCvShare...
-    cv_ptr = cv_bridge::toCvCopy(img_msg);
+    cv_constptr = cv_bridge::toCvShare(img_msg);
```
@ruffsl this commit msg caught my eye, bcs I think I tried toCvShare, but last time I tried there was something in the code modifying the images (probably in the frontend, which I am heavily refactoring now) , just FYI

ruffsl replied:
Ok, I'll revert that commit for the time being. Is there a ticket I could watch on that refactoring upstream?

ToniRV replied:
Ah, unfortunately, these tickets are in the private repo...

ruffsl replied:
@ToniRV , with the latest pushes, is this still an issue?

ToniRV replied:
I'm afraid it still is. I am working on this on another branch that hasn't been merged yet. Sorry for the delay!

ruffsl committed:
Use type alias instead of using namespace

ruffsl committed:
Adding TODO comments

ruffsl committed:
Rename nodes to components

ruffsl committed:
Register components

ruffsl committed:
Export ament dependencies

ruffsl committed:
Register the component with class_loader

ruffsl committed:
Comment out launch install for now

ruffsl committed:
Cleanup CMake

ruffsl committed:
Ignore .vscode

ruffsl committed:
Declare frame id params

ruffsl committed:
Add depends for output interfaces

ruffsl committed:
Stage backend interface

ruffsl committed:
Uncrustify

ruffsl committed:
Enable glog checks

ruffsl committed:
Add publishState for odometry

ruffsl committed:
Use alias for message types

ruffsl committed:
Init Google Logging

ruffsl committed:
Stage virtual class changes
TODO: solve bad weak pointer at runtime

ruffsl committed:
Stage launchfiles, params, flags

ruffsl committed:
Use SharedPtr for node in constructors
to avoid bad_weak_ptr runtime error

ruffsl committed:
Register Backend Output Callback

ruffsl committed:
Correct frame_id params

ruffsl committed:
Sping pipline by starting async thread

ruffsl committed:
Set is rectified to true

ruffsl committed:
Enable run parallel
to avoid manual spinging pipline

ruffsl committed:
Hack arround data race with camera params

ruffsl committed:
Include upstrea, patch of launch in workspace
to Fix parsing of cmd line arguments in XML and yaml file

ruffsl commented on kimera_vio_ros/src/interfaces/imu_interface.cpp:
```diff
@@ -0,0 +1,67 @@
+    imu_accgyr(4) = imu_msg->angular_velocity.y;
+    imu_accgyr(5) = imu_msg->angular_velocity.z;
+
+    this->imu_single_callback_(VIO::ImuMeasurement(timestamp, imu_accgyr));
```
@ToniRV , I'm still having some issues with `No IMU data from last frame timestamp:` when using `parallel_run=true`. Any suggestions other than setting parallel_run to false. I'd like to keep `pipline_.spin()` in it's own async thread.

```
$ ros2 launch kimera_vio_ros kimera_ros_realsense.launch.xml 
[INFO] [launch]: All log files can be found below /home/ruffsl/.ros/log/2020-03-02-18-45-16-443174-dox-12704
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [kimera_vio_ros-1]: process started with pid [12715]
[kimera_vio_ros-1] I0302 18:45:16.720127 12715 CameraParams.cpp:31] Parsing camera parameters for: left_cam
[kimera_vio_ros-1] I0302 18:45:16.787691 12715 CameraParams.cpp:31] Parsing camera parameters for: right_cam
[kimera_vio_ros-1] I0302 18:45:16.788328 12715 DataProviderInterface.cpp:96] Running dataset between frame 50 and frame 10000
[kimera_vio_ros-1] I0302 18:45:16.796392 12715 RegularVioBackEnd.cpp:113] Using Regular VIO backend.
[kimera_vio_ros-1] I0302 18:45:17.805269 12734 PipelineModule.h:163] Module: Data Provider - Spinning.
[kimera_vio_ros-1] W0302 18:45:17.817641 12734 DataProviderModule.cpp:97] No IMU data from last frame timestamp: 1583203517581242112 to timestamp: 1583203517614594304
[kimera_vio_ros-1] W0302 18:45:17.818171 12734 DataProviderModule.cpp:97] No IMU data from last frame timestamp: 1583203517581242112 to timestamp: 1583203517614594304
[kimera_vio_ros-1] W0302 18:45:17.818605 12734 DataProviderModule.cpp:97] No IMU data from last frame timestamp: 1583203517581242112 to timestamp: 1583203517614594304
...
```

ruffsl replied:
I think I was encountering an error with an uninitialized  '_P' camera param given my setting of `rectified=true` (which should be true given that's what the realsense publishes), but found out I was using an incorrect nominalBaseline before that was failing the rectification setup that would normally initialized  '_P'. However, now I'm getting a rather strange issue with the pipline setup with Eigen:

```
$ ros2 launch kimera_vio_ros kimera_ros_realsense.launch.xml 
[INFO] [launch]: All log files can be found below /home/ruffsl/.ros/log/2020-03-02-21-18-24-670948-dox-3734
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [kimera_vio_ros-1]: process started with pid [3745]
[kimera_vio_ros-1] I0302 21:18:24.941282  3745 CameraParams.cpp:31] Parsing camera parameters for: left_cam
[kimera_vio_ros-1] I0302 21:18:24.993412  3745 CameraParams.cpp:31] Parsing camera parameters for: right_cam
[kimera_vio_ros-1] I0302 21:18:24.994200  3745 DataProviderInterface.cpp:96] Running dataset between frame 50 and frame 10000
[kimera_vio_ros-1] I0302 21:18:25.002403  3745 RegularVioBackEnd.cpp:113] Using Regular VIO backend.
[kimera_vio_ros-1] I0302 21:18:25.010413  3759 PipelineModule.h:163] Module: Data Provider - Spinning.
[kimera_vio_ros-1] I0302 21:18:25.198184  3759 Pipeline.cpp:780] Frontend launched (parallel_run set to 1).
[kimera_vio_ros-1] I0302 21:18:25.198218  3775 PipelineModule.h:163] Module: VioFrontEnd - Spinning.
[kimera_vio_ros-1] I0302 21:18:25.198231  3759 Pipeline.cpp:511] ------------------- Initialize Pipeline with frame k = 1--------------------
[kimera_vio_ros-1] stereo_vio_node: /usr/include/eigen3/Eigen/src/Core/PlainObjectBase.h:374: void Eigen::PlainObjectBase<Derived>::resizeLike(const Eigen::EigenBase<OtherDerived>&) [with OtherDerived = Eigen::Matrix<double, 6, -1>; Derived = Eigen::Matrix<double, 6, 1>]: Assertion `other.rows() == 1 || other.cols() == 1' failed.
```

ruffsl replied:
Resolved with https://github.com/MIT-SPARK/Kimera-VIO/issues/37

ruffsl committed:
Revert data race hack

ruffsl committed:
Correct nominalBaseline with realsense config

ruffsl committed:
Use timer or async to spin pipline given parallel

ruffsl committed:
Disable visualizer

ruffsl committed:
Uncrustify

ruffsl committed:
Give pipline timer own callback group

ruffsl committed:
Comment out handle_shutdown_

ruffsl committed:
Clean up use of Node::SharedPtr

ruffsl committed:
Add missing optical_flow_predictor_type param

ruffsl committed:
Enable lcd

ruffsl committed:
Simplify shutdown

ruffsl committed:
Match use of CHECK with ROS1 wrapper

ruffsl committed:
Stage additions for publishTimeHorizonPointCloud

ruffsl committed:
Move tf broadcaster to base and add listener

ruffsl committed:
Update branch

ruffsl committed:
Rename launchfile

ruffsl committed:
Pin gtsam to last known working commit
head of devel branch seems to have broken an added VIO test
src/MIT-SPARK/Kimera-VIO/tests/testLogger.cpp:415:31: error: no matching function for call to ‘gtsam::Rot3::Random(boost::random::mt19937&)’
       gtsam::Rot3::Random(rng_), gtsam::Point3::Random());
                               ^

ruffsl committed:
Add simple demo example for testing

ruffsl commented on install/underlay.repos:
```diff
@@ -2,7 +2,7 @@ repositories:
     type: git
     url: https://github.com/borglab/gtsam.git
-    version: develop
+    version: 26a5a42a38375d539162f8094ab273e032e2146c
```
@ToniRV I think the latest changes in the `develop` branch from gtam have broken the build for Kimera_VIO, see 1859c10b2d463dbdde3f689de2a97f3865407622 . Do you know of the most recent commit suported by kimera. Gtsam hasn't had any recent releases, and the commit I pinned is from around January.

ToniRV replied:
The latest commit I have tried in GTSAM is: 
> (last tested but not pushed with commit `ee069286b447ff58b809423cc77c777a02abdfe5`)
> Previously tested commits: `0c3e05f375c03c5ff5218e708db416b38f4113c8`

Besides a minor compilation error due to the use of `std` instead of `boost` on the tests (https://github.com/borglab/gtsam/pull/233/files#), I haven't noticed any problem.

ruffsl replied:
Looks like things are building fine after updating to latest head as of 5df50d6

ruffsl committed:
Always import entire wrapper wrok space

ruffsl committed:
Patch the hardcoded path_to_vocab
until kimera_vio fixes it's install rules

ruffsl committed:
Publish static TF so base_link to imu is identity

Related:
https://github.com/MIT-SPARK/Kimera-VIO/issues/41

ruffsl committed:
Add note on TODO
and clean up launch file

ruffsl committed:
Omit note on un-reproducible error

ruffsl committed:
Rename vio pipline pointer

ruffsl committed:
Add vio_params_ field

ruffsl committed:
Use make_unique

ruffsl committed:
Declare vio params folder path

ruffsl committed:
Use UniquePtr for vio_pipeline_

ruffsl committed:
Add geometry utils

ruffsl committed:
Use poseToMsgTF from geometry utils

ruffsl committed:
Rename tf buffer

ruffsl committed:
Add msgCamInfoToCameraParams to ImageInterface

ruffsl committed:
void camera_info_cb to StereoInterface

ruffsl committed:
Update launchfiles

ruffsl committed:
Fix topics

ruffsl committed:
Add param use_camera_info

ruffsl committed:
Uncrustify

ruffsl committed:
Clean up old launch params

ruffsl committed:
Update realsense params

ruffsl committed:
Disable use_camera_info for now

ruffsl committed:
Update .repos

ruffsl committed:
Adjust plumb_bob to expected radtan format

ruffsl committed:
Subscribe to images after camera info received

ruffsl committed:
Re-enable use_camera_info

ruffsl committed:
Uncrustify

ruffsl commented on kimera_vio_ros/src/interfaces/base_interface.cpp:
```diff
@@ -0,0 +1,74 @@
+  params_folder_ = node_->declare_parameter(
+    "params_folder", "");
+  CHECK(!params_folder_.empty());
+  vio_params_ = std::make_shared<VIO::VioParams>(
```
@ToniRV After the recent changes to the front end, it looks like vio_params_ has been made separate from VIO::DataProviderInterface class. This might make it easier to work with when using stuff like camera_info topics to init the extrinsic and intrinsic camera parameters, so thanks. 

However, it looks like it seeks a number of yaml files with hardcoded file names, like the left and right camera yamls, but also some new files that where never added to the realsense data set params folder that would include a reference to the default hyperparameters for that device:

* https://github.com/MIT-SPARK/Kimera-VIO/tree/83e12698a5803dcb6e6111dbed25b7ef59c878bb/params/RealSense

Could the reference params folder for the RealSense be updated to re-sync with the recent refactoring updtream?

ruffsl commented on kimera_vio_ros/src/interfaces/stereo_interface.cpp:
```diff
@@ -0,0 +1,152 @@
+    // RCLCPP_INFO(this->get_logger(), "Did you forget to register the right frame callback?");
+
+    left_frame_callback_(VIO::make_unique<VIO::Frame>(
+        frame_count_, timestamp_left, left_cam_info, readRosImage(left_msg)));
```
@ToniRV I saw a TODO comment in the KimeraVIO fronted when debugging https://github.com/MIT-SPARK/Kimera-VIO-ROS2/pull/5/commits/ec3586292209a953e671963e710ac84a6878f089 about the idea of removing the camera info argument from the frame callback.

https://github.com/ruffsl/Kimera-VIO/blob/e5e799c631b7ae3b8d3afac323996a381297348b/src/dataprovider/EurocDataProvider.cpp#L147-L149

 I just wanted to mention that keeping it might be nice for setups with changing extrinsic and intrinsic camera parameters.

E.g. When the camera has automatic focus or zoom, or when the camera to imu tranform is non-static, like with a robot's or drone articulated head with pan tilt control, but the IMU in fixed in place at the center of the chassis. These changes could still be picked up and synchronized with the data per-frame by consciously subscribing to camera_info and tf topics.

ruffsl commented on kimera_vio_ros/src/interfaces/imu_interface.cpp:
```diff
@@ -0,0 +1,67 @@
+    imu_accgyr(2) = imu_msg->linear_acceleration.z;
+    imu_accgyr(3) = imu_msg->angular_velocity.x;
+    imu_accgyr(4) = imu_msg->angular_velocity.y;
+    imu_accgyr(5) = imu_msg->angular_velocity.z;
```
@ToniRV Would 9DoF IMUs ever be supported? E.g. where a magnetometer's orientation reading can be used. Does GTSAM even have an IMU factor that can consider absolute orientation?

ruffsl commented on kimera_vio_ros/src/interfaces/base_interface.cpp:
```diff
@@ -0,0 +1,74 @@
+
+BaseInterface::BaseInterface(
+  rclcpp::Node::SharedPtr & node)
+: VIO::DataProviderInterface(),
```
@ToniRV , Instead of inheriting from the `VIO::DataProviderInterface` class, It might be cleaner to replicate what was changed in the ROS1 wrapper and create a `data_provider_` filed:

* https://github.com/MIT-SPARK/Kimera-VIO-ROS/blob/ec5f3d4325612b4526f81e3314bda1c38836ae59/src/KimeraVioRos.cpp#L36

However, sterio and imu child interfaces here wouldn't then be able to access the private vio callback functions:

* https://github.com/MIT-SPARK/Kimera-VIO/blob/d677a3dedad2da605c84a4b7f51154cd7e38a98d/include/kimera-vio/dataprovider/DataProviderInterface.h#L77-L83

I tried creating a protected child class `VIO::DataProviderInterface`, and listing `BaseInterface` as a protected friend, but that workaround didn't seem to compile. It might of just have been my error in playing with the C++ classes.

ruffsl commented on kimera_vio_ros/src/interfaces/backend_interface.cpp:
```diff
@@ -0,0 +1,227 @@
+      odometry_msg.twist
+      .covariance[i * static_cast<int>(
+          sqrt(odometry_msg.twist.covariance.size())) +
+        j] = vel_cov_body(i, j);
```
@ToniRV , In relation to https://github.com/MIT-SPARK/Kimera-VIO/issues/41 , I think if there was a way to convert twist.linear and twist.covariance of the odometry_msg (or velocity_body) into a different reference frame, the then the wrapper could do the abstraction of allowing the imu frame and base link frames to have a non-identity displacement transform. This would also be handy if the transform between the IMU and base-link was non-static, like on sterio-imu pan tilt unit. But it would be nice to keep such reference frame transform functaions down in the kimera-vio library for all wrappers to share.

ruffsl commented on kimera_vio_ros/src/interfaces/image_interface.cpp:
```diff
@@ -0,0 +1,106 @@
+    // Kimera-VIO terms the plumb bob dist. model the as radtan.
+    cam_params->distortion_model_ = "radtan";
+    // Kimera-VIO can't take a 6th order radial distortion term.
+    CHECK_EQ(cam_info->d.size(), 5);
```
@ToniRV it would be nice if Kimera-VIO understood the `plumb_bob` type to be the same as `radtan`:
* https://github.com/ros/common_msgs/blob/ea6c65e8846d9873b8d6e44435c52fb43d6093a3/sensor_msgs/include/sensor_msgs/distortion_models.h#L45
* https://github.com/ros/common_msgs/blob/ea6c65e8846d9873b8d6e44435c52fb43d6093a3/sensor_msgs/msg/CameraInfo.msg#L58

And also if Kimera-VIO library determined for itself whether to use or not the 6th order radial distortion term.

Also would be cool if Kimera-VIO checked if all the distortion terms where zero, so no rectification would be necessary. Not sure if `cv::initUndistortRectifyMap` internally optimizes for that. Lots of times, ROS image topics will already be rectified to avoid each downstream image consuming node from having to duplicate the rectification. 

https://github.com/ros-perception/vision_opencv/blob/0bbebb73bea65eca856e732b919439853a04986b/image_geometry/src/pinhole_camera_model.cpp#L133-L148

ruffsl committed:
Fix camera info received logic

ruffsl committed:
Clean up callback names

ruffsl commented on kimera_vio_ros/src/stereo_vio_node.cpp:
```diff
@@ -0,0 +1,23 @@
+  auto g_args = rclcpp::init_and_remove_ros_arguments(argc, argv);
+  int g_argc = g_args.size();
+  // Initialize Google's flags library.
+  google::ParseCommandLineFlags(&g_argc, &argv, true);
```
@ToniRV , when using composed nodes, I don't think we'd be able to acquire the CLI args for the containing process that holds all the nodes. Thus we'll need a way to pass any current run time args for google flag/log using ROS parameters instead. Perhaps we could a ROS parameter for passing a directories path containing google flag files, like with the param path for yaml files. Or would there be would be a way to combine the two?

ruffsl committed:
Clean up subscription changes

ruffsl committed:
Remove implicit opencv dependency
captured by  cv_bridge
